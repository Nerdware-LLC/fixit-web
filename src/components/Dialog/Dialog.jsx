import React from "react";
import MuiDialog from "@mui/material/Dialog";
import DialogTitle from "@mui/material/DialogTitle";
import DialogContent from "@mui/material/DialogContent";
import DialogContentText from "@mui/material/DialogContentText";
import DialogActions from "@mui/material/DialogActions";
import { SlideTransition } from "../Transitions/SlideTransition";
import { Button } from "../Button";
import { bool, string, func, shape, styleType } from "../../types";

// Docs: https://mui.com/components/dialogs/

export const Dialog = ({
  isVisible,
  title,
  message,
  handleAccept,
  handleCancel,
  acceptLabel = "OK",
  cancelLabel = "CANCEL",
  styles
}) => (
  <MuiDialog
    open={isVisible}
    onClose={handleCancel}
    aria-labelledby={"dialog-title"}
    aria-describedby={"dialog-message"}
    TransitionComponent={SlideTransition}
    style={styles?.container}
    fullWidth
  >
    <DialogTitle id={"dialog-title"} style={styles?.title}>
      {title}
    </DialogTitle>
    <DialogContent dividers>
      <DialogContentText id={"dialog-message"} style={styles?.message}>
        {message}
      </DialogContentText>
    </DialogContent>
    <DialogActions>
      {handleCancel && (
        <Button label={cancelLabel} onClick={handleCancel} color={"primary"} />
      )}
      <Button label={acceptLabel} onClick={handleAccept} color={"primary"} />
    </DialogActions>
  </MuiDialog>
);

Dialog.propTypes = {
  isVisible: bool.isRequired,
  title: string.isRequired,
  message: string.isRequired,
  handleAccept: func,
  handleCancel: func,
  acceptLabel: string,
  cancelLabel: string,
  styles: shape({
    container: styleType,
    title: styleType,
    message: styleType
  })
};
