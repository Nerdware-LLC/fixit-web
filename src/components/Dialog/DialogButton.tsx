import { styled } from "@mui/material/styles";
import Button, { type ButtonProps } from "@mui/material/Button";

export const DialogButton = (buttonProps: ButtonProps) => (
  <StyledDialogButton className="dialog-button" {...buttonProps}></StyledDialogButton>
);

const StyledDialogButton = styled(Button)({
  height: "2.5rem",
  minWidth: "5.5rem",
  padding: "0.75rem",
  paddingBottom: "0.5rem",
  fontSize: "1rem",
  lineHeight: "1rem",
  fontWeight: "bold",
  backgroundColor: "primary.main",

  "@media (max-width: 550px)": {
    fontWeight: 500,
    maxWidth: "40%",
    whiteSpace: "pre-line"
  },

  "@media (max-width: 375px)": {
    height: "3.5rem",
    fontSize: "0.9rem",
    padding: 0,
    paddingTop: "0.25rem"
  }
});
