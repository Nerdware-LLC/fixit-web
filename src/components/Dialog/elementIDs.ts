/**
 * IDs of `Dialog` components (used for aria props).
 */
export const dialogElementIDs = {
  title: "dialog-title",
  message: "dialog-message",
} as const;
