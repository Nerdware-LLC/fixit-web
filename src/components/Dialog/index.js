export { Dialog } from "./Dialog";
export { useDialog } from "./useDialog";
