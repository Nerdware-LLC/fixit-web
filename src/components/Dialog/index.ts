export * from "./Dialog";
export * from "./useDialog";
export * from "./elementIDs";
