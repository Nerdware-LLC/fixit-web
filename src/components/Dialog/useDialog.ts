import { useState } from "react";

export const useDialog = (initIsVisible = false) => {
  const [isDialogVisible, setIsDialogVisible] = useState(!!initIsVisible);

  const openDialog = () => setIsDialogVisible(true);
  const closeDialog = () => setIsDialogVisible(false);

  return {
    isDialogVisible,
    openDialog,
    closeDialog,
  };
};

export type UseDialogHookProps = ReturnType<typeof useDialog>;
