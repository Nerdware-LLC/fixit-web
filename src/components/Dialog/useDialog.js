import { useState } from "react";
import { Dialog } from "./Dialog";

export const useDialog = (initIsVisible = false) => {
  const [isDialogVisible, setIsDialogVisible] = useState(!!initIsVisible);

  const openDialog = () => setIsDialogVisible(true);
  const closeDialog = () => setIsDialogVisible(false);

  return { Dialog, isDialogVisible, openDialog, closeDialog };
};
