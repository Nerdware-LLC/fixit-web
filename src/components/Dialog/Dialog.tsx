import MuiDialog from "@mui/material/Dialog";
import DialogTitle from "@mui/material/DialogTitle";
import DialogContent from "@mui/material/DialogContent";
import DialogContentText from "@mui/material/DialogContentText";
import DialogActions from "@mui/material/DialogActions";
import { SlideTransition } from "@components/Transitions/SlideTransition";
import { DialogButton } from "./DialogButton";
import { useDialog } from "./useDialog";

/**
 * Docs: https://mui.com/components/dialogs/
 */
export const Dialog = ({
  isVisible,
  title,
  message,
  handleAccept,
  handleCancel,
  acceptLabel = "OK",
  cancelLabel = "CANCEL",
  ...containerProps
}: {
  isVisible: boolean;
  title: React.ReactNode;
  message: React.ReactNode;
  handleAccept?: React.MouseEventHandler<HTMLButtonElement>;
  handleCancel?: React.MouseEventHandler<HTMLButtonElement>;
  acceptLabel?: React.ReactNode;
  cancelLabel?: React.ReactNode;
} & Omit<React.ComponentProps<typeof MuiDialog>, "open">) => (
  <MuiDialog
    open={isVisible}
    onClose={handleCancel}
    aria-labelledby="dialog-title"
    aria-describedby="dialog-message"
    TransitionComponent={SlideTransition}
    fullWidth
    {...containerProps}
  >
    <DialogTitle color="secondary">{title}</DialogTitle>
    <DialogContent dividers>
      {typeof message === "string" ? (
        <DialogContentText color="action" style={{ whiteSpace: "pre-line" }}>
          {message}
        </DialogContentText>
      ) : (
        message
      )}
    </DialogContent>
    <DialogActions
      style={{
        padding: "1rem 0.5rem",
        justifyContent: "space-evenly"
      }}
    >
      {handleCancel && (
        <DialogButton onClick={handleCancel} variant="outlined">
          {cancelLabel}
        </DialogButton>
      )}
      <DialogButton onClick={handleAccept}>{acceptLabel}</DialogButton>
    </DialogActions>
  </MuiDialog>
);

Dialog.use = useDialog;
