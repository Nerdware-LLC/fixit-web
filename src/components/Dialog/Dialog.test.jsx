import ReactDOM from "react-dom";
import { render } from "@testing-library/react";
import { Dialog } from "./Dialog";

beforeAll(() => {
  // Mock ReactDOM.createPortal for components built on top of Portal.
  // (react-test-renderer does not support this OOB)
  ReactDOM.createPortal = jest.fn((element, node) => {
    return element;
  });
});

afterEach(() => {
  ReactDOM.createPortal.mockClear();
});

/* NOTE: for Portal'ed components, we have to use @testing-library/react,
react-test-renderer doesn't support Portals.  */

it("renders correctly", () => {
  expect(
    render(
      <div>
        <Dialog
          isVisible={true}
          title="Test Dialog Title"
          message="This is a test dialog message."
        />
      </div>
    ).baseElement
  ).toMatchSnapshot();
});
