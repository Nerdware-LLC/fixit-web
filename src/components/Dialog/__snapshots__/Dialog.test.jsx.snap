// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div />
  </div>
  <div
    class="MuiModal-root MuiDialog-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-describedby="dialog-message"
        aria-labelledby="dialog-title"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth css-tlc64q-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
          id="dialog-title"
        >
          Test Dialog Title
        </h2>
        <div
          class="MuiDialogContent-root MuiDialogContent-dividers css-1t4vnk2-MuiDialogContent-root"
        >
          <p
            class="MuiDialogContentText-root MuiTypography-root MuiTypography-body1 css-qfso29-MuiTypography-root-MuiDialogContentText-root"
            id="dialog-message"
          >
            This is a test dialog message.
          </p>
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
        >
          <button
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-1p1iq82-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            OK
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
